components/configable/config_derive/src/lib.rs

```diff
@@ -0,0 +1,156 @@
+            #update_fn
+            #diff_fn
+        }
+    })
+}
+
+fn get_struct_fields(
+    span: proc_macro2::Span,
+    data: &Data,
+) -> std::result::Result<&Punctuated<Field, Comma>, Error> {
+    if let Data::Struct(DataStruct {
+        fields: Fields::Named(FieldsNamed { ref named, .. }),
+        ..
+    }) = data
+    {
+        Ok(named)
+    } else {
+        Err(Error::new(
+            span,
+            "expect derive Configable on struct with named fields!",
+        ))
+    }
+}
+
+fn update(
+    fields: &Punctuated<Field, Comma>,
+    creat_name: &Ident,
+) -> Result<proc_macro2::TokenStream> {
+    let incoming = Ident::new("incoming", proc_macro2::Span::call_site());
+    let mut update_fields = Vec::with_capacity(fields.len());
+    for field in fields {
+        let (not_support, submodule) = get_attrs(&field.attrs)?;
+        if not_support || field.ident.is_none() {
+            continue;
+        }
+        let name = field.ident.as_ref().unwrap();
+        let name_lit = LitStr::new(&format!("{}", name), name.span());
+        let f = if submodule {
+            quote! {
+                if let Some(#creat_name::ConfigValue::Module(v)) = #incoming.remove(#name_lit) {
+                    #creat_name::Configable::update(&mut self.#name, v);
+                }
+            }
+        } else {
+            quote! {
+                if let Some(v) = #incoming.remove(#name_lit) {
+                    self.#name = v.into();
+                }
+            }
+        };
+        update_fields.push(f);
+    }
+    Ok(quote! {
+        fn update(&mut self, mut #incoming: #creat_name::ConfigChange) {
+            #(#update_fields)*
+        }
+    })
+}
+
+fn diff(fields: &Punctuated<Field, Comma>, creat_name: &Ident) -> Result<proc_macro2::TokenStream> {
+    let diff_ident = Ident::new("diff_ident", proc_macro2::Span::call_site());
+    let incoming = Ident::new("incoming", proc_macro2::Span::call_site());
+    let mut diff_fields = Vec::with_capacity(fields.len());
+    for field in fields {
+        let (not_support, submodule) = get_attrs(&field.attrs)?;
+        if not_support || field.ident.is_none() {
+            continue;
+        }
+        let name = field.ident.as_ref().unwrap();
+        let name_lit = LitStr::new(&format!("{}", name), name.span());
+        let f = if submodule {
+            quote! {
+                {
+                    let diff = #creat_name::Configable::diff(&self.#name, &#incoming.#name);
+                    if diff.len() != 0 {
+                        #diff_ident.insert(#name_lit.to_owned(), #creat_name::ConfigValue::from(diff));
+                    }
+                }
+            }
+        } else {
+            quote! {
+                if self.#name != #incoming.#name {
+                    #diff_ident.insert(#name_lit.to_owned(), #creat_name::ConfigValue::from(#incoming.#name.clone()));
+                }
+            }
+        };
+        diff_fields.push(f);
```

Comment:
Instead of expanding code at compile-time, could you fold them into a static array and iterate them at run-time? This helps reducing compiling time and binary size.